File: structural/Java/Decorator.java
interface Widget {

	public void draw();

	public String getDescription();

}

class Window implements Widget {

	@Override
	public void draw() {
		System.out.println("Drawing the Window");
	}

	@Override
	public String getDescription() {
		return "Window";
	}

}

class WindowDecorator extends Window {

	private Widget windowToBeDecorated;

	public WindowDecorator(Widget windowToBeDecorated) {
		this.windowToBeDecorated = windowToBeDecorated;
	}

	@Override
	public void draw() {
		this.windowToBeDecorated.draw(); // Delegation
	}

	public String getDescription() {
		return this.windowToBeDecorated.getDescription(); // Delegation
	}

}

class VerticalScrollBarDecorator extends WindowDecorator {

	public VerticalScrollBarDecorator(Widget windowToBeDecorated) {
		super(windowToBeDecorated);
	}

	@Override
	public void draw() {
		super.draw();
		this.drawVerticalScrollBar();
	}

	@Override
	public String getDescription() {
		return super.getDescription() + ", including vertical scrollbars";
	}

	private void drawVerticalScrollBar() {
		System.out.println("Drawing Vertical ScrollBar");
	}

}

class HorizontalScrollBarDecorator extends WindowDecorator {

	public HorizontalScrollBarDecorator(Widget windowToBeDecorated) {
		super(windowToBeDecorated);
	}

	@Override
	public void draw() {
		super.draw();
		this.drawHorizontalScrollBar();
	}

	@Override
	public String getDescription() {
		return super.getDescription() + ", including horizontal scrollbars";
	}

	private void drawHorizontalScrollBar() {
		System.out.println("Drawing Horizontal ScrollBar");
	}

}

public class Main {

	public static void main(String[] args) {
		System.out.println(":: Decorator ::\n");

		Window decoratedWindow = new HorizontalScrollBarDecorator(new VerticalScrollBarDecorator(new Window()));

		System.out.println(decoratedWindow.getDescription());
	}

}

File: structural/javascript/Decorator.js
// Interface
class Widget {

  constructor() {
    if (this.constructor === Widget) {
      throw new TypeError('Cannot construct Abstract instances directly');
    }

    if (this.draw === Widget.prototype.draw) {
      throw new TypeError('Must override method "draw()"');
    }

    if (this.getDescription === Widget.prototype.getDescription) {
      throw new TypeError('Must override method "getDescription()"');
    }
  }

  draw() {
    throw new TypeError('Do not call abstract method "draw()" from child.');
  }

  getDescription() {
    throw new TypeError('Do not call abstract method "getDescription()" from child.');
  }

}

class Window extends Widget {

  // Implements abstract parent's method
  draw() {
    console.log('Drawing the Window');
  }

  // Implements abstract parent's method
  getDescription() {
    return 'Window';
  }

}

class WindowDecorator extends Window {

  constructor(windowToBeDecorated) {
    super();
    this.windowToBeDecorated = windowToBeDecorated;
  }

  // override
  draw() {
    this.windowToBeDecorated.draw(); // Delegation
  }

  // override
  getDescription() {
    return this.windowToBeDecorated.getDescription(); // Delegation
  }

}

class VerticalScrollBarDecorator extends WindowDecorator {

  constructor(windowToBeDecorated) {
    super(windowToBeDecorated);
  }

  // override
  draw() {
    super.draw();
    this.drawVerticalScrollBar();
  }

  // override
  getDescription() {
    return super.getDescription() + ', including vertical scrollbars';
  }

  drawVerticalScollBar() {
    console.log('Drawing Vertical ScrollBar');
  }

}

class HorizontalScrollBarDecorator extends WindowDecorator {

  constructor(windowToBeDecorated) {
    super(windowToBeDecorated);
  }

  // override
  draw() {
    super.draw();
    this.drawHorizontalScrollBar();
  }

  // override
  getDescription() {
    return super.getDescription() + ', including horizontal scrollbar';
  }

  drawHorizontalScrollBar() {
    console.log('Drawing Horizontal ScrollBar');
  }

}

console.log(':: Decorator ::\n');

const decoratedWindow = new HorizontalScrollBarDecorator(new VerticalScrollBarDecorator(new Window()));

console.log(decoratedWindow.getDescription());